packages/eslint-plugin/src/rules/restrict-template-expressions.ts

```diff
@@ -12,15 +13,44 @@ import {
   isTypeNeverType,
 } from '../util';
 
+type OptionTester = (
+  type: Type,
+  checker: TypeChecker,
+  recursivelyCheckType: (type: Type) => boolean,
+) => boolean;
+
+const testTypeFlag =
+  (flagsToCheck: TypeFlags): OptionTester =>
+  type =>
+    isTypeFlagSet(type, flagsToCheck);
+
+const optionTesters = (
+  [
+    ['Any', isTypeAnyType],
+    [
+      'Array',
+      (type, checker, recursivelyCheckType): boolean =>
+        checker.isArrayType(type) &&
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        recursivelyCheckType(type.getNumberIndexType()!),
+    ],
+    // eslint-disable-next-line @typescript-eslint/internal/prefer-ast-types-enum
```

Comment:
@JoshuaKGoldberg - Originally this rule's purpose was to prevent code like `node.type === 'string'`.
Given we now have the `no-unsafe-enum-comparison` rule - do you think this internal rule still has value? WDYT about deleting it? (we already have nuec turned on via the recommended config)

Reply:
Yeah I'd be in favor of that. It wouldn't be the first example of an internal lint rule ending up as a subset of a general good one.